desc/objectives/_generic.py

```diff
@@ -257,13 +257,14 @@ def build(self, use_jit=False, verbose=1):
         self._dim_f = jax.eval_shape(self._fun, thing.params_dict).size
 
         # check that fun is linear
+        params, params_tree = tree_flatten(thing.params_dict)
+        for param in params:
+            param += np.random.rand(param.size) * 10
+        params = tree_unflatten(params_tree, params)
         J1 = jax.jacrev(self._fun)(thing.params_dict)
-        params = thing.params_dict.copy()
-        for key, value in params.items():
-            params[key] = value + np.random.rand(value.size) * 10
         J2 = jax.jacrev(self._fun)(params)
-        for key in J1.keys():
-            assert np.all(J1[key] == J2[key]), "Function must be linear!"
+        for j1, j2 in zip(tree_leaves(J1), tree_leaves(J2)):
```

Comment:
why was this change needed?

Reply:
This is technically unrelated to the signed curvature, but I threw it in this PR (sorry). This change is necessary to allow `LinearObjectiveFromUser` to work with an `OptimizableCollection` like a `CoilSet`.

Reply:
is this tested anywhere in this PR?

Reply:
The new functionality is not tested in this PR, no. But there are existing tests for `LinearObjectiveFromUser` that still pass.

Reply:
see #1130 for a test of this new functionality